File: benchmarks/29-graph-ops/graph_ops.m
(* Experiment 29 -- Graph operations.
   MEASURES src/graph/ (19 files): construction, degree sequences, connected
   components, shortest paths, spanning trees.

   NOT COVERED BY THE EXISTING CORPUS.  Experiment 12 measured graph ALGORITHMS
   written by hand out of gathers and segmented sums, deliberately avoiding a
   graph type, to isolate the gather.  This experiment measures the opposite: the
   src/graph/ builtins themselves, which nothing has ever timed.

   Baseline is networkx, which is pure Python -- so a Mathilda win here is a
   weaker claim than a win against a compiled library, and the .py docstring says
   so.  What the row set is really for is finding builtins that are absent or
   accidentally quadratic. *)

Get["../harness.m"];
Get["../data.m"];

require[{"Graph", "EdgeList", "VertexList", "VertexDegree", "AdjacencyMatrix",
         "ConnectedComponents", "FindShortestPath", "GraphDistance",
         "ConnectedGraphQ", "FindSpanningTree", "VertexCount", "EdgeCount"}];

(* A deterministic sparse graph: a cycle plus chords, 20000 vertices. *)
nv = 20000;
edges = Join[Table[i -> Mod[i, nv] + 1, {i, 1, nv}],
             Table[i -> Mod[3 i, nv] + 1, {i, 1, nv}]];
g = Graph[edges];

bench["Graph construction, 40000 edges", Graph[edges], 1];
check["Graph construction, 40000 edges", Length[edges]];

bench["VertexDegree, 20000 vertices", VertexDegree[g];];
check["VertexDegree, 20000 vertices",
  Total[VertexDegree[Graph[{1 -> 2, 2 -> 3, 3 -> 1}]]]];

bench["EdgeCount", EdgeCount[g];];
check["EdgeCount", EdgeCount[Graph[{1 -> 2, 2 -> 3, 3 -> 1}]]];

benchIf["ConnectedComponents", "ConnectedComponents",
  ConnectedComponents[g], 1];
(* Mathematica counts differently on a DIRECTED graph, so the check uses an
   undirected two-component graph where all three systems must agree. *)
checkIf["ConnectedComponents", "ConnectedComponents",
  Length[ConnectedComponents[Graph[{1 <-> 2, 3 <-> 4}]]]];

benchIf["GraphDistance from vertex 1", "GraphDistance",
  GraphDistance[g, 1], 1];
checkIf["GraphDistance from vertex 1", "GraphDistance",
  GraphDistance[Graph[{1 -> 2, 2 -> 3}], 1, 3]];

benchIf["FindShortestPath 1 to 10000", "FindShortestPath",
  FindShortestPath[g, 1, 10000], 1];
checkIf["FindShortestPath 1 to 10000", "FindShortestPath",
  Length[FindShortestPath[Graph[{1 -> 2, 2 -> 3}], 1, 3]]];

benchIf["AdjacencyMatrix, 20000 vertices", "AdjacencyMatrix",
  AdjacencyMatrix[g], 1];
checkIf["AdjacencyMatrix, 20000 vertices", "AdjacencyMatrix",
  Total[Flatten[AdjacencyMatrix[Graph[{1 -> 2, 2 -> 3, 3 -> 1}]]]]];

File: benchmarks/29-graph-ops/graph_ops.py
#!/usr/bin/env python3
"""Experiment 29 -- Graph operations (networkx column).

Same seven kernels as ``graph_ops.m``, same order and sizes.

NOT A COMPILED BASELINE.  networkx is pure Python, so a Mathilda win here is a
weaker claim than a win against scipy or numpy.  What this row set is really for
is finding src/graph/ builtins that are ABSENT or accidentally quadratic --
nothing in the existing corpus has ever timed them (experiment 12 deliberately
avoided a graph type in order to isolate the gather).

The graph is deterministic -- a directed cycle plus 3i chords on 20000 vertices --
so all three systems build the same object and the checks are comparable.
"""

import sys, os; sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

import networkx as nx

from harness import bench, check, require

require(["networkx", "networkx:DiGraph", "networkx:shortest_path"])

nv = 20000
edges = ([(i, i % nv + 1) for i in range(1, nv + 1)]
         + [(i, (3 * i) % nv + 1) for i in range(1, nv + 1)])


def build():
    g = nx.DiGraph()
    g.add_edges_from(edges)
    return g


g = build()
tri = nx.DiGraph([(1, 2), (2, 3), (3, 1)])

bench("Graph construction, 40000 edges", build, reps=1)
check("Graph construction, 40000 edges", len(edges))

# Wolfram's VertexDegree is total degree (in + out) for a directed graph.
bench("VertexDegree, 20000 vertices", lambda: [d for _, d in g.degree()])
check("VertexDegree, 20000 vertices", sum(d for _, d in tri.degree()))

bench("EdgeCount", lambda: g.number_of_edges())
check("EdgeCount", tri.number_of_edges())

# Wolfram's ConnectedComponents on a directed graph gives WEAKLY connected ones.
bench("ConnectedComponents",
      lambda: list(nx.weakly_connected_components(g)), reps=1)
check("ConnectedComponents",
      nx.number_connected_components(nx.Graph([(1, 2), (3, 4)])))

bench("GraphDistance from vertex 1",
      lambda: nx.single_source_shortest_path_length(g, 1), reps=1)
check("GraphDistance from vertex 1",
      nx.shortest_path_length(nx.DiGraph([(1, 2), (2, 3)]), 1, 3))

bench("FindShortestPath 1 to 10000",
      lambda: nx.shortest_path(g, 1, 10000), reps=1)
check("FindShortestPath 1 to 10000",
      len(nx.shortest_path(nx.DiGraph([(1, 2), (2, 3)]), 1, 3)))

bench("AdjacencyMatrix, 20000 vertices",
      lambda: nx.adjacency_matrix(g), reps=1)
check("AdjacencyMatrix, 20000 vertices",
      int(nx.adjacency_matrix(tri).sum()))
